esmvalcore/_recipe.py

```diff
@@ -1017,6 +1017,12 @@ def _get_preprocessor_products(variables, profile, order, ancestor_products,
     for product in products | multimodel_products | ensemble_products:
         product.check()
 
+        # Ensure that attributes start_year and end_year are always available
+        # for all products
+        start_year, end_year = _parse_period(product.attributes['timerange'])
+        product.attributes['start_year'] = int(str(start_year[0:4]))
+        product.attributes['end_year'] = int(str(end_year[0:4]))
```

Comment:
Yes, this will overwrite all existing attributes, but `start_year` and `end_year` should always reflect the `timerange`, so I don't think this is a problem.

Reply:
Maybe this can be done in `initialize_variables`, that way the products would not be modified.

Reply:
Yes, that could certainly be done, but I intentionally added this here. It's not the first time that products are missing `start_year` and `end_year` (e.g., #1485), and by adding it here we make sure that every product that comes out of the preprocessor has this attributes.

Regarding overwriting pre-existing attributes: `start_year` and `end_year` are removed for every variable at the [beginning of the preprocessor chain](https://github.com/ESMValGroup/ESMValCore/blob/5f6145cb73fc04cd61230617bc346022935e08a2/esmvalcore/_data_finder.py#L161-L172) and added back at various points using the same code that is used here, so this is definitely not a problem.

Reply:
Yes this was done because during the review of the changes, it was considered that having the three tags was too crowded. But it turns out that having the start and end year is necessary for diagnostics. So afterall, maybe it would be good to ensure the tags are present from the very beginning. But whatever it does need to be covered in this PR

However I tried running the test adding wildcards in the timerange  and it fails, which does not happen when working with regular variables.

Reply:
You mean you ran the new test `test_derive_contains_start_end_year` with a wildcard in `timerange` and it failed? Can you post the traceback?

Reply:
Yes, I set `timerange: '2000/*` and got this:

```
tests/integration/test_recipe.py:281: in get_recipe
    recipe = read_recipe_file(str(recipe_file), cfg)
esmvalcore/_recipe.py:74: in read_recipe_file
    return Recipe(raw_recipe,
esmvalcore/_recipe.py:1252: in __init__
    self.tasks = self.initialize_tasks() if initialize_tasks else None
esmvalcore/_recipe.py:1861: in initialize_tasks
    tasks = self._create_tasks()
esmvalcore/_recipe.py:1836: in _create_tasks
    new_tasks, failed = self._create_preprocessor_tasks(
esmvalcore/_recipe.py:1796: in _create_preprocessor_tasks
    task = _get_preprocessor_task(
esmvalcore/_recipe.py:1195: in _get_preprocessor_task
    derive_input = _get_derive_input_variables(variables, config_user)
esmvalcore/_recipe.py:1160: in _get_derive_input_variables
    files = _get_input_files(var, config_user)[0]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

variable = {'alias': 'GFDL-CM3', 'dataset': 'GFDL-CM3', 'derive': True, 'diagnostic': 'diagnostic_name', ...}
config_user = {'auxiliary_data_dir': '/tmp/pytest-of-sloosvel/pytest-1686/test_derive_contains_start_end0/extra_data', 'check_level': <CheckLevels.DEFAULT: 3>, 'compress_netcdf': False, 'config_developer_file': None, ...}

    def _get_input_files(variable, config_user):
        """Get the input files for a single dataset (locally and via download)."""
        if variable['frequency'] != 'fx':
            start_year, end_year = _parse_period(variable['timerange'])
    
            start_year = int(str(start_year[0:4]))
>           end_year = int(str(end_year[0:4]))
E           ValueError: invalid literal for int() with base 10: '*'

esmvalcore/_recipe.py:583: ValueError

```

Reply:
But this is a different issue, right? This exact same error should also appear in the current `main` branch. It looks very much like #1516 but for derived variables instead of automatic downloads.

I guess using wildcards for derived variables is also not trivial since it can depend on multiple input variables...

Reply:
Right, then I guess this PR does what it should do, feel free to merge!

Reply:
> I agree that this does not need to be addressed here, but I think it is a bit different from the issue of automatic downloading. With automatic downloading, it is not possible (right now) to determine what might be available. For derived variables, the timerange should be determined from the constituent variables, right? So perhaps we should have a separate ticket for that?

Doing that as we speak

Reply:
See #1550